modules/standard/Random.chpl

```diff
@@ -222,14 +291,332 @@ module Random {
     randNums.permutation(arr);
   }
 
+  /*
+    Use a new :record:`randomStream` to produce a random permutation of an
+    array's domain. The values ``d.dim(0).low..d.dim(0).high`` will appear
+    exactly once in the array in a pseudo-random order.
+
+    :arg arr: The array to store the permutation in
+    :arg seed: The seed to use when initializing a ``randomStream``
+  */
+  @unstable("'permutation' is unstable and subject to change")
+  proc permutation(ref arr: [?d] ?t, seed: int)
+    where isCoercible(d.idxType, t) && is1DRectangularDomain(d) do
+  {
+    var rs = new randomStream(d.idxType, seed);
+    rs.permutation(arr);
+  }
+
+  /*
+    Use a new :record:`randomStream` to produce a random permutation of an
+    array's domain. The values ``d.dim(0).low..d.dim(0).high`` will appear
+    exactly once in the array in a pseudo-random order.
+
+    :arg arr: The array to store the permutation in
+  */
+  @unstable("'permutation' is unstable and subject to change")
+  proc permutation(ref arr: [?d] ?t)
+    where isCoercible(d.idxType, t) && is1DRectangularDomain(d) do
+  {
+    var rs = new randomStream(d.idxType);
+    rs.permutation(arr);
+  }
+
+  @chpldoc.nodoc
+  proc permutation(ref arr: [], seed: int) {
+    compilerError("'permutation' only supports 1D rectangular arrays whose domain idxType is compatible with eltType");
+  }
+
   @chpldoc.nodoc
-  proc makeRandomStream(type eltType,
-                        seed: int(64) = SeedGenerator.oddCurrentTime,
-                        param parSafe: bool = true,
-                        param algorithm = defaultRNG) {
-    compilerWarning("makeRandomStream is deprecated - " +
-                    "please use createRandomStream instead");
-    return createRandomStream(eltType, seed, parSafe, algorithm);
+  proc permutation(ref arr: []) {
+        compilerError("'permutation' only supports 1D rectangular arrays whose domain idxType is compatible with eltType");
+  }
+
+  record randomStream: writeSerializable {
+    /*
+      Specifies the type of value generated by the random stream. Currently
+      numeric and bool types are supported.
+    */
+    type eltType;
+
+    /* The seed value for the PCG random number generator */
+    const seed: int;
+
+    @chpldoc.nodoc
+    var pcg: shared PCGRandomStreamInternal(eltType, false);
```

Comment:
IMO it's very important that this be a record storing the state here rather than a class. The reason is, I want things like assigning to different `randomStream`s across locales to copy the state rather than refer to shared state. I think this behavior is important in the context of making `randomStream` not have locking.

To resolve this comment, I'd recommend changing `class PCGRandomStreamInternal` into `record pcgRandomStreamImpl` and then introducing a class to wrap that for use in deprecated code `class DeprecatedPcgRandomStream`. (Or, perhaps this is a case where a deprecation warning + `compilerError` call is acceptable, a in which case you could simplify this a bit and might not need to create `class DeprecatedPcgRandomStream` at all).

Reply:
An alternative approach would be to declare the state here like it is done in `PCGRandomStreamInternal` & to factor all of the method bodies out into private non-method routines that accept the state as a 1st argument and that both call `class PCGRandomStreamInternal` and this record call.

Reply:
Also I think we will need a new test that locks in the behavior that copying a `randomStream` to a new variable and getting random values out of it will not change the behavior of the original when it is used.